File: ncov.py
import re,requests,json,pprint,time,datetime
from time import strftime

pattern=re.compile(r'^try{sinajp_15844213244528328543098388435\((.*?)\);}catch\(e\){};')
lasttimes='00:00:00'
year = datetime.datetime.now().strftime('%Y')

res=requests.get('https://gwpre.sina.cn/ncp/foreign?_=1584421324452&callback=sinajp_15844213244528328543098388435')
match=pattern.search(res.text)
if match:
    obj=json.loads(match.group(1))
    resultObj=obj['result']
    times=resultObj['times']  # 截止时间
    timesMatch=re.search(r'截至(\d{2})月(\d{2})日(\d{2})时(\d{2})分',times)
    if timesMatch:
        times=year+'年'+timesMatch.group(1)+'月'+timesMatch.group(2)+'日 '+timesMatch.group(3)+':'+timesMatch.group(4)
        timeStr = year+timesMatch.group(1)+timesMatch.group(2)+'-'+timesMatch.group(3)+timesMatch.group(4)
    
        lasttimes=times
        totalObj=resultObj['total']
        certain=totalObj['certain'] # 累计确诊
        die=totalObj['die']   # 死亡
        recure=totalObj['recure'] # 治愈
        certain_inc=totalObj['certain_inc'] # 确诊增加
        die_inc=totalObj['die_inc'] # 死亡增加
        recure_inc=totalObj['recure_inc'] # 治愈增加
        # 各国数据列表
        worldlistArr=resultObj['worldlist']
        worldlistArr.sort(key=lambda x: int(x.get('conNum','0')),reverse=True)
        
        f=open('./docs/others/'+ timeStr +'.md','a',encoding='utf-8')
        f.writelines('\n# '+times+'\n')
        f.writelines('感染国家总数：'+str(len(worldlistArr))+'\n')
        f.writelines('```\n累计确诊：'+certain.rjust(10,' ')+' 较昨日：'+certain_inc+'\n'+'累计死亡：'+die.rjust(10,' ')+' 较昨日：'+die_inc+'\n'+'累计治愈：'+recure.rjust(10,' ')+' 较昨日：'+recure_inc+'\n```\n')

        # 写入echarts横向柱状图
        countrys = []
        conadds = []
        conNums = []
        deathadds = []
        cureNums = []
        deathNums = []

        countrys_second = []
        conadds_second = []
        conNums_second = []
        deathadds_second = []
        cureNums_second = []
        deathNums_second = []

        countrys_third = []
        conadds_third = []
        conNums_third = []
        deathadds_third = []
        cureNums_third = []
        deathNums_third = []

        countrys_last = []
        conadds_last = []
        conNums_last = []
        deathadds_last = []
        cureNums_last = []
        deathNums_last = []

        worldlistArr.reverse()

        for item in worldlistArr[190:]: 
          if item['name'] =='中国':
            continue
          countrys.append(item['name'])
          conadds.append((item['conadd']))
          conNums.append((item['conNum']))
          deathadds.append((item['deathadd']))
          cureNums.append((item['cureNum']))
          deathNums.append((item['deathNum']))

        for item in worldlistArr[180:190]: 
          if item['name'] =='中国':
            continue
          countrys_second.append(item['name'])
          conadds_second.append((item['conadd']))
          conNums_second.append((item['conNum']))
          deathadds_second.append((item['deathadd']))
          cureNums_second.append((item['cureNum']))
          deathNums_second.append((item['deathNum']))

        for item in worldlistArr[170:180]: 
          if item['name'] =='中国':
            continue
          countrys_third.append(item['name'])
          conadds_third.append((item['conadd']))
          conNums_third.append((item['conNum']))
          deathadds_third.append((item['deathadd']))
          cureNums_third.append((item['cureNum']))
          deathNums_third.append((item['deathNum']))

        for item in worldlistArr[0:170]: 
          if item['name'] =='中国':
            continue
          countrys_last.append(item['name'])
          conadds_last.append((item['conadd']))
          conNums_last.append((item['conNum']))
          deathadds_last.append((item['deathadd']))
          cureNums_last.append((item['cureNum']))
          deathNums_last.append((item['deathNum']))

        print(countrys)
        print(conNums)
        print(cureNums)
        print(deathNums)

        f.writelines('\n<div id="main" style="width:100%;height:800px;margin-bottom:10px;"></div>\n')
        f.writelines('\n<div id="second" style="width:100%;height:1000px;margin-bottom:10px;"></div>\n')
        f.writelines('\n<div id="third" style="width:100%;height:1000px;margin-bottom:10px;"></div>\n')
        f.writelines('\n<div id="last" style="width:100%;height:3000px;"></div>\n')
        f.writelines('\n')
        f.writelines('<script>\n')
        f.writelines('import * as echarts from "echarts";\n')
        f.writelines('export default {\n')
        f.writelines('mounted() {\n')
        f.writelines('this.chart = echarts.init(document.getElementById("main"),"dark")\n')
        f.writelines('this.secondChart = echarts.init(document.getElementById("second"),"dark")\n')
        f.writelines('this.thirdChart = echarts.init(document.getElementById("third"),"dark")\n')
        f.writelines('this.lastChart = echarts.init(document.getElementById("last"),"dark")\n')

        f.writelines('var option = {\n')
        f.writelines('tooltip: {trigger: "axis",axisPointer: {type: "shadow"}},\n')
        f.writelines('legend: {},\n')
        f.writelines('grid: {left: "3%",right: "4%",bottom: "3%",containLabel: true},\n')
        f.writelines('xAxis: {type: "value"},\n')
        f.writelines('yAxis: {type: "category",data: [\n')
        for country in countrys: 
          f.writelines("'"+  country +"',")
        f.writelines(']},\n')
        f.writelines('series: [\n')
        f.writelines('{name: "新增确诊",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(conadds) +']},')
        f.writelines('{name: "累计确诊",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(conNums) +']},')
        f.writelines('{name: "新增死亡",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(deathadds) +']},')
        f.writelines('{name: "累计死亡",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(deathNums) +']},')
        f.writelines('{name: "累计治愈",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(cureNums) +']},')
        f.writelines(']}\n')
        f.writelines('this.chart.setOption(option);\n')

        f.writelines('var secondOption = {\n')
        f.writelines('tooltip: {trigger: "axis",axisPointer: {type: "shadow"}},\n')
        f.writelines('legend: {},\n')
        f.writelines('grid: {left: "3%",right: "4%",bottom: "3%",containLabel: true},\n')
        f.writelines('xAxis: {type: "value"},\n')
        f.writelines('yAxis: {type: "category",data: [\n')
        for country in countrys_second: 
          f.writelines("'"+  country +"',")
        f.writelines(']},\n')
        f.writelines('series: [\n')
        f.writelines('{name: "新增确诊",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(conadds_second) +']},')
        f.writelines('{name: "累计确诊",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(conNums_second) +']},')
        f.writelines('{name: "新增死亡",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(deathadds_second) +']},')
        f.writelines('{name: "累计死亡",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(deathNums_second) +']},')
        f.writelines('{name: "累计治愈",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(cureNums_second) +']},')
        f.writelines(']}\n')
        f.writelines('this.secondChart.setOption(secondOption);\n')

        f.writelines('var thirdOption = {\n')
        f.writelines('tooltip: {trigger: "axis",axisPointer: {type: "shadow"}},\n')
        f.writelines('legend: {},\n')
        f.writelines('grid: {left: "3%",right: "4%",bottom: "3%",containLabel: true},\n')
        f.writelines('xAxis: {type: "value"},\n')
        f.writelines('yAxis: {type: "category",data: [\n')
        for country in countrys_third: 
          f.writelines("'"+  country +"',")
        f.writelines(']},\n')
        f.writelines('series: [\n')
        f.writelines('{name: "新增确诊",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(conadds_third) +']},')
        f.writelines('{name: "累计确诊",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(conNums_third) +']},')
        f.writelines('{name: "新增死亡",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(deathadds_third) +']},')
        f.writelines('{name: "累计死亡",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(deathNums_third) +']},')
        f.writelines('{name: "累计治愈",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(cureNums_third) +']},')
        f.writelines(']}\n')
        f.writelines('this.thirdChart.setOption(thirdOption);\n')

        f.writelines('var lastOption = {\n')
        f.writelines('tooltip: {trigger: "axis",axisPointer: {type: "shadow"}},\n')
        f.writelines('legend: {},\n')
        f.writelines('grid: {left: "3%",right: "4%",bottom: "3%",containLabel: true},\n')
        f.writelines('xAxis: {type: "value"},\n')
        f.writelines('yAxis: {type: "category",data: [\n')
        for country in countrys_last: 
          f.writelines("'"+  country +"',")
        f.writelines(']},\n')
        f.writelines('series: [\n')
        f.writelines('{name: "新增确诊",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(conadds_last) +']},')
        f.writelines('{name: "累计确诊",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(conNums_last) +']},')
        f.writelines('{name: "新增死亡",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(deathadds_last) +']},')
        f.writelines('{name: "累计死亡",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(deathNums_last) +']},')
        f.writelines('{name: "累计治愈",type: "bar",stack: "total",label: {show: true},emphasis:{focus: "series"},data: ['+ ','.join(cureNums_last) +']},')
        f.writelines(']}\n')
        f.writelines('this.lastChart.setOption(lastOption);\n')

        f.writelines('}};')
        f.writelines('</script>\n')
        f.writelines('\n')

        # 写入表格
        f.writelines('|国家|新增确诊|累计确诊|新增死亡|累计死亡|累计治愈|'+'\n')
        f.writelines('|:--:|---:|---:|---:|---:|---:|'+'\n')

        worldlistArr.reverse()
        
        for countryObj in worldlistArr:
            name=countryObj['name'] # 国家
            if name=='中国':
                continue
            conadd=countryObj['conadd'] # 新增确诊
            conNum=countryObj['conNum'] # 累计确诊
            deathadd=countryObj['deathadd'] # 新增死亡
            deathNum=countryObj['deathNum'] # 累计死亡
            cureNum=countryObj['cureNum'] # 累计治愈
            f.writelines('|'+name+'|'+conadd+'|'+conNum+'|'+deathadd+'|'+deathNum+'|'+cureNum+'|\n')
        f.close()

    

File: jsSrc/ncov.js
'use strict'
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod }
  }
Object.defineProperty(exports, '__esModule', { value: true })
const fs_1 = __importDefault(require('fs'))
const axios_1 = __importDefault(require('axios'))
const ncov_1 = require('./configs/ncov')
/**
 * @func getNcovText
 * @param {string} url
 * @returns {object}
 * @desc 获取疫情数据
 */
const getNcovText = async (url) => {
  const res = await axios_1.default.get(url)
  // console.log(res.data)
  const r1 = res.data.replace('try{sinajp_15844213244528328543098388435(', '')
  const r2 = r1.replace(');}catch(e){};', '')
  const r3 = JSON.parse(r2)
  const { times, total, worldlist } = r3.result
  return {
    times,
    total,
    worldlist
  }
}
/**
 * @func writeMdWithContent
 * @param {string} timeStr
 * @param {string} content
 * @desc 写入md文件并更新路由
 */
const writeMdWithContent = (timeStr, content) => {
  const writePath = `${ncov_1.rootPath}/docs/others/${timeStr}.md`
  fs_1.default.writeFileSync(writePath, content, 'utf-8')
  console.log(`${timeStr}.md created.`)
  setTimeout(() => {
    const filesList = (0, ncov_1.readFileList)(ncov_1.mdPath)
    console.log(ncov_1.mdPath)
    console.log(filesList)
    console.log('读取文件目录生成路由---')
    const writeFileList = (path) => {
      try {
        const obj = [
          {
            text: '历史记录',
            link: `/${ncov_1.base}/${ncov_1.othersNcovCategory}.md`
          }
        ]
        fs_1.default.writeFileSync(path, JSON.stringify(obj))
        console.log('写入路由到JSON文件---')
      } catch (error) {
        console.log(error)
      }
    }
    writeFileList(ncov_1.jsonFilePath)
    ;(0, ncov_1.writeOthersNcovCategory)(filesList)
  }, 500)
}
const createContent = async () => {
  const res = await getNcovText(ncov_1.SINA_URL)
  const { times, total, worldlist } = res
  // console.log(res)
  const certain = total['certain'] // 累计确诊
  const die = total['die'] // 死亡
  const recure = total['recure'] // 治愈
  const certain_inc = total['certain_inc'] // 确诊增加
  const die_inc = total['die_inc'] // 死亡增加
  const recure_inc = total['recure_inc'] // 治愈增加
  const worldlistArr = worldlist
    .sort((a, b) => {
      return b.conNum - a.conNum
    })
    .filter((item) => {
      return !ncov_1.Exclude_Countrys.includes(item.name)
    })
    .reverse()
  console.log(worldlistArr.length)
  // console.log(times)
  const year = new Date().getFullYear()
  const month = times.slice(2, 4)
  const day = times.slice(5, 7)
  const hour = times.slice(8, 10)
  const minute = times.slice(11, 13)
  const timeStr = `${year}${month}${day}-${hour}${minute}` // '20220904-0720'
  const title = `${year}年${month}月${day}日 ${hour}:${minute}`
  const countrys = []
  const conadds = []
  const conNums = []
  const deathadds = []
  const cureNums = []
  const deathNums = []
  const countrys_second = []
  const conadds_second = []
  const conNums_second = []
  const deathadds_second = []
  const cureNums_second = []
  const deathNums_second = []
  const countrys_third = []
  const conadds_third = []
  const conNums_third = []
  const deathadds_third = []
  const cureNums_third = []
  const deathNums_third = []
  const countrys_last = []
  const conadds_last = []
  const conNums_last = []
  const deathadds_last = []
  const cureNums_last = []
  const deathNums_last = []
  // console.log(worldlistArr.slice(189))
  worldlistArr.slice(189).forEach((item) => {
    countrys.push(item['name'])
    conadds.push(item['conadd'])
    conNums.push(item['conNum'])
    deathadds.push(item['deathadd'])
    cureNums.push(item['cureNum'])
    deathNums.push(item['deathNum'])
  })
  worldlistArr.slice(179, 189).forEach((item) => {
    countrys_second.push(item['name'])
    conadds_second.push(item['conadd'])
    conNums_second.push(item['conNum'])
    deathadds_second.push(item['deathadd'])
    cureNums_second.push(item['cureNum'])
    deathNums_second.push(item['deathNum'])
  })
  worldlistArr.slice(169, 179).forEach((item) => {
    countrys_third.push(item['name'])
    conadds_third.push(item['conadd'])
    conNums_third.push(item['conNum'])
    deathadds_third.push(item['deathadd'])
    cureNums_third.push(item['cureNum'])
    deathNums_third.push(item['deathNum'])
  })
  worldlistArr.slice(0, 169).forEach((item) => {
    countrys_last.push(item['name'])
    conadds_last.push(item['conadd'])
    conNums_last.push(item['conNum'])
    deathadds_last.push(item['deathadd'])
    cureNums_last.push(item['cureNum'])
    deathNums_last.push(item['deathNum'])
  })
  const content = `
# ${title}
感染国家总数：${worldlist.length}
\`\`\`
累计确诊： ${certain} 较昨日：${certain_inc}
累计死亡：   ${die} 较昨日：${die_inc}
累计治愈： ${recure} 较昨日：${recure_inc}
\`\`\`
<div id="main" style="width:100%;height:800px;margin-bottom:10px;"></div>
<div id="second" style="width:100%;height:1000px;margin-bottom:10px;"></div>
<div id="third" style="width:100%;height:1000px;margin-bottom:10px;"></div>
<div id="last" style="width:100%;height:3000px;"></div>

<script>
import * as echarts from "echarts";
export default {
  mounted () {
    this.chart = echarts.init(document.getElementById("main"), "dark")
    this.secondChart = echarts.init(document.getElementById("second"), "dark")
    this.thirdChart = echarts.init(document.getElementById("third"), "dark")
    this.lastChart = echarts.init(document.getElementById("last"), "dark")
    var option = {
      tooltip: { trigger: "axis", axisPointer: { type: "shadow" } },
      legend: {},
      grid: { left: "3%", right: "4%", bottom: "3%", containLabel: true },
      xAxis: { type: "value" },
      yAxis: {
        type: "category", data: [${countrys
          .map((x) => {
            return `"${x}",`
          })
          .join('')}]
      },
      series: [
        { name: "新增确诊", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${conadds
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "累计确诊", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${conNums
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "新增死亡", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${deathadds
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "累计死亡", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${deathNums
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "累计治愈", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${cureNums
          .map((x) => {
            return `${x},`
          })
          .join('')}] },]
    }
    this.chart.setOption(option);
    var secondOption = {
      tooltip: { trigger: "axis", axisPointer: { type: "shadow" } },
      legend: {},
      grid: { left: "3%", right: "4%", bottom: "3%", containLabel: true },
      xAxis: { type: "value" },
      yAxis: {
        type: "category", data: [${countrys_second
          .map((x) => {
            return `"${x}",`
          })
          .join('')}]
      },
      series: [
        { name: "新增确诊", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${conadds_second
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "累计确诊", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${conNums_second
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "新增死亡", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${deathadds_second
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "累计死亡", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${deathNums_second
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "累计治愈", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${cureNums_second
          .map((x) => {
            return `${x},`
          })
          .join('')}] },]
    }
    this.secondChart.setOption(secondOption);
    var thirdOption = {
      tooltip: { trigger: "axis", axisPointer: { type: "shadow" } },
      legend: {},
      grid: { left: "3%", right: "4%", bottom: "3%", containLabel: true },
      xAxis: { type: "value" },
      yAxis: {
        type: "category", data: [${countrys_third
          .map((x) => {
            return `"${x}",`
          })
          .join('')}]
      },
      series: [
        { name: "新增确诊", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${conadds_third
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "累计确诊", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${conNums_third
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "新增死亡", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${deathadds_third
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "累计死亡", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${deathNums_third
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "累计治愈", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${cureNums_third
          .map((x) => {
            return `${x},`
          })
          .join('')}] },]
    }
    this.thirdChart.setOption(thirdOption);
    var lastOption = {
      tooltip: { trigger: "axis", axisPointer: { type: "shadow" } },
      legend: {},
      grid: { left: "3%", right: "4%", bottom: "3%", containLabel: true },
      xAxis: { type: "value" },
      yAxis: {
        type: "category", data: [${countrys_last
          .map((x) => {
            return `"${x.replace(/\s*/g, '')}",`
          })
          .join('')}]
      },
      series: [
        { name: "新增确诊", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${conadds_last
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "累计确诊", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${conNums_last
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "新增死亡", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${deathadds_last
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "累计死亡", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${deathNums_last
          .map((x) => {
            return `${x},`
          })
          .join('')}] }, 
        { name: "累计治愈", type: "bar", stack: "total", label: { show: true }, emphasis: { focus: "series" }, data: [${cureNums_last
          .map((x) => {
            return `${x},`
          })
          .join('')}] },]
    }
    this.lastChart.setOption(lastOption);

    window.onresize = () => {
      this.chart.resize()
      this.secondChart.resize()
      this.thirdChart.resize()
      this.lastChart.resize()
    }
  }
};
</script>

|国家|新增确诊|累计确诊|新增死亡|累计死亡|累计治愈|
|:--:|---:|---:|---:|---:|---:|
${worldlistArr
  .reverse()
  .map((item) => {
    return `|${item.name}|${item.conadd}|${item.conNum}|${item.deathadd}|${item.deathNum}|${item.cureNum}|\n`
  })
  .join('')}
`
  writeMdWithContent(timeStr, content)
}
createContent()
